node/beacon/src/authority.rs

```diff
@@ -81,51 +90,106 @@ impl Authority {
+            self.processed_blocks = processed_blocks;
+            let thresholds = guard.get_threshold();
+            self.thresholds = thresholds;
+            let accepted_authorities = guard.get_accepted_authorities();
+            self.accepted_authorities = accepted_authorities;
+        }
+    }
+
+    fn prune_slot<V: Clone>(&self, data: &HashMap<Slot, V>) -> HashMap<Slot, V> {
+        data.iter()
+            .filter(|(&k, &_)| self.slot_to_epoch(k) >= self.current_epoch)
+            .map(|(k, v)| (k.clone(), v.clone()))
+            .collect()
+    }
+
+    fn prune_epoch<V: Clone>(&self, data: &HashMap<Epoch, V>) -> HashMap<Epoch, V> {
+        data.iter()
+            .filter(|(&k, &_)| k >= self.current_epoch)
+            .map(|(k, v)| (k.clone(), v.clone()))
+            .collect()
+    }
+
+    fn write_to_storage(&self) {
+        // prune old epochs
+        let proposals = self.prune_slot(&self.proposals);
+        let participation = self.prune_slot(&self.participation);
+        let processed_blocks = self.prune_epoch(&self.processed_blocks);
+        let thresholds = self.prune_epoch(&self.thresholds);
+        let accepted_authorities = self.prune_slot(&self.accepted_authorities);
+        let mut guard = self.storage.write().expect(POISONED_LOCK_ERR);
+        guard.set_proposal(&proposals);
+        guard.set_participation(&participation);
+        guard.set_processed_blocks(&processed_blocks);
+        guard.set_threshold(&thresholds);
+        guard.set_accepted_authorities(&accepted_authorities);
+    }
+
     /// Initializes authorities from the config and the past blocks in the beaconchain.
-    pub fn new(authority_config: AuthorityConfig, blockchain: &BeaconBlockChain) -> Self {
+    pub fn new(
+        authority_config: AuthorityConfig,
+        blockchain: &BeaconBlockChain,
+        storage: Arc<RwLock<BeaconChainStorage>>
+    ) -> Self {
         // TODO: cache authorities in the Storage, to not need to process the whole chain.
         let mut result = Self {
             authority_config,
             proposals: HashMap::new(),
             participation: HashMap::new(),
             processed_blocks: HashMap::new(),
+            current_epoch: 0,
             thresholds: HashMap::new(),
             accepted_authorities: HashMap::new(),
+            storage,
         };
-        // Initial authorities operate for the first two epochs.
-        let (accepted_authorities, threshold) = result.compute_threshold_accepted(
-            &CryptoHash::default(),
-            result.authority_config.initial_proposals.to_vec(),
-            vec![],
-        );
-        let mut slot = 0;
-        for epoch in 0..=1 {
-            result.thresholds.insert(epoch, threshold);
-            for slot_auth in &accepted_authorities {
-                slot += 1;
-                result.accepted_authorities.insert(slot, slot_auth.to_vec());
+        result.read_from_storage();
```

Comment:
what about current_epoch? 
doesn't seem like it's calculated if read from storage?

Reply:
I don't think it is necessary to calculate that. `self.current_epoch` is only used for pruning and is  only updated when we need to write to storage.

Reply:
Let's then not have it as a field, instead pass it to `write_to_storage(epoch)` and to prune methods.
Ppl will use it if it's available and it will lead to weird bugs.

Reply:
Fixed